### PR TITLE
Cassandra 18129 livemigration trunk

### PR DESCRIPTION
**Pull request by aweisberg**

### CODE REVIEW

File: accord-core/src/main/java/accord/local/PreLoadContext.java [hunk 1]
```diff
@@ -162,4 +162,6 @@ static PreLoadContext empty()
     {
         return contextFor(null, Collections.emptyList(), Keys.EMPTY);
     }
+
+    PreLoadContext EMPTY_PRELOADCONTEXT = empty();
```
Comment (bdeggleston) [inline accord-core/src/main/java/accord/local/PreLoadContext.java]: unused

File: accord-core/src/main/java/accord/messages/ReadData.java [hunk 2]
```diff
@@ -68,10 +99,17 @@ protected ReadData(TxnId txnId, Participants<?> readScope, long waitForEpoch)
         this.waitForEpoch = waitForEpoch;
     }
 
+    public ReadType kind()
```
Comment (bdeggleston) [inline accord-core/src/main/java/accord/messages/ReadData.java]: can this just be made abstract?

File: accord-core/src/main/java/accord/local/SafeCommandStore.java [hunk 3]
```diff
@@ -213,10 +212,15 @@ protected SafeCommand get(TxnId txnId)
      * Visits keys first and then ranges, both in ascending order.
      * Within each key or range visits TxnId in ascending order of queried timestamp.
      */
-    public abstract <T> T mapReduce(Seekables<?, ?> keys, Ranges slice,
+    public abstract <T> T mapReduceWithTerminate(Seekables<?, ?> keysOrRanges, Ranges slice,
```
Comment (bdeggleston) [inline accord-core/src/main/java/accord/local/SafeCommandStore.java:215]: same comment wrt removing `withTerminate` from the method name

File: accord-core/src/main/java/accord/coordinate/CoordinateSyncPoint.java [hunk 4]
```diff
@@ -19,74 +19,114 @@
...
+    private static <S extends Seekables<?, ?>> AsyncResult<CoordinateSyncPoint<S>> coordinate(Node node, Kind kind, S keysOrRanges, boolean async)
     {
-        TxnId txnId = node.nextTxnId(kind, ranges.domain());
-        return node.withEpoch(txnId.epoch(), () -> {
-            FullRangeRoute route = (FullRangeRoute) node.computeRoute(txnId, ranges);
-            CoordinateSyncPoint coordinate = new CoordinateSyncPoint(node, txnId, node.agent().emptyTxn(kind, ranges), route, ranges);
-            coordinate.start();
-            return coordinate;
-        }).beginAsResult();
+        checkState(kind == Kind.SyncPoint || kind == ExclusiveSyncPoint);
```
Comment (bdeggleston) [inline accord-core/src/main/java/accord/coordinate/CoordinateSyncPoint.java]: I think we'd want this to be a checkArgument?

**Review by bdeggleston — CHANGES_REQUESTED** [review]

File: accord-core/src/main/java/accord/local/PreLoadContext.java (hunk 1 above)
Comment (aweisberg) [reply accord-core/src/main/java/accord/local/PreLoadContext.java]: Huh, certainly don't remember why I created it, but I made it used by having `empty()` return the global empty preload context.

**Review by bdeggleston — APPROVED** [review]
+1

**Pull request merged**